### ISSUE TITLE
disk usage of app running with docker image 

### ISSUE THREAD

**Issue by scottgai** [cloudfoundry/garden-runc-release#81]
Hi there! Thanks for taking time to open an Issue.
Please fill out as much of the following information as you can.

# Description
Pushing an app with docker image would fail if default disk quota (1G) is used since the docker image size is about 1.3G. 

```
[root@mdw ~]# CF_DOCKER_PASSWORD=xxxx cf push img1 --docker-username xxxx  --docker-image scottgai/sgai-repo:img1 scottgai -t 600
   2018-06-25T23:29:06.20-0400 [CELL/0] OUT Cell 21c1ae94-0f86-4dc5-83ec-7b76dd2dba58 creating container for instance da9af175-a04a-4079-49f7-89d5
   2018-06-25T23:29:07.88-0400 [CELL/0] ERR Cell 21c1ae94-0f86-4dc5-83ec-7b76dd2dba58 failed to create container for instance da9af175-a04a-4079-49f7-89d5: running image plugin create: making image: creating image: applying disk limits: disk limit is smaller than volume size
   2018-06-25T23:29:07.88-0400 [CELL/0] ERR : exit status 1
   2018-06-25T23:29:07.88-0400 [CELL/0] OUT Cell 21c1ae94-0f86-4dc5-83ec-7b76dd2dba58 destroying container for instance da9af175-a04a-4079-49f7-89d5
   2018-06-25T23:29:07.88-0400 [CELL/0] OUT Cell 21c1ae94-0f86-4dc5-83ec-7b76dd2dba58 successfully destroyed container for instance da9af175-a04a-4079-49f7-89d5
   2018-06-25T23:29:07.90-0400 [API/0] OUT Process has crashed with type: "web"
   2018-06-25T23:29:07.90-0400 [API/0] OUT App instance exited with guid 6b783e7e-4a21-4d52-8a6b-a99b7147ed22 payload: {"instance"=>"da9af175-a04a-4079-49f7-89d5", "index"=>0, "reason"=>"CRASHED", "exit_description"=>"failed to initialize container: running image plugin create: making image: creating image: applying disk limits: disk limit is smaller than volume size\n: exit status 1", "crash_count"=>3, "crash_timestamp"=>1529983747883330324, "version"=>"a6bc244b-30cb-49f5-8f22-0778392d1b73"}
```

`cf push` with disk quota 2G was successful. However `cf app <app name>` only shows very small value of disk usage ("52K of 2G" in following test. seems only show size of platform overhead excluding docker image size)
```
[root@mdw ~]# CF_DOCKER_PASSWORD=xxxx cf push img1 --docker-username xxxx  --docker-image scottgai/sgai-repo:img1 scottgai -t 600 -k 2G
Using docker repository password from environment variable CF_DOCKER_PASSWORD.
Pushing app img1 to org org1 / space dev as admin...
Getting app info...
Updating app with these attributes...
  name:                   img1
  docker image:           scottgai/sgai-repo:img1
  docker username:        scottgai
  command:                /docker-entrypoint.sh httpd -k start -D FOREGROUND
- disk quota:             1G
+ disk quota:             2G
  health check timeout:   600
  health check type:      port
  instances:              1
  memory:                 1G
  stack:                  cflinuxfs2
  routes:
    img1.apps-03.haas-50.pez.pivotal.io

Updating app img1...
Mapping routes...

Stopping app...

Waiting for app to start...

name:              img1
requested state:   started
instances:         1/1
usage:             1G x 1 instances
routes:            img1.apps-03.haas-50.pez.pivotal.io
last uploaded:     Mon 25 Jun 23:27:32 EDT 2018
stack:             cflinuxfs2
docker image:      scottgai/sgai-repo:img1
start command:     /docker-entrypoint.sh httpd -k start -D FOREGROUND

     state     since                  cpu    memory      disk        details
#0   running   2018-06-26T03:48:35Z   0.0%   25M of 1G   52K of 2G
```
Is the disk usage being displayed correctly?

# Environment

* garden-runc-release version: 1.13.1
* IaaS: vSphere
* Stemcell version: 3468.46
* Kernel version: 4.4.0-127-generic

If known, provide the resolution to the issue here.

**Comment by scottgai** [comment]
can someone comment if this is an issue with displaying disk usage of app with docker image?

Thanks,
Scott

**Comment by Callisto13** [comment]
i wasn't aware that garden (or groot) sets a default disk size. @julz is this happening elsewhere?

**Comment by scottgai** [comment]
Can anyone give some update to this issue?

**Comment by Callisto13** [comment]
Hi @scottgai,
It seems we may have seen this before: https://github.com/cloudfoundry/cloud_controller_ng/issues/1106.
Would you mind confirming which Diego version is deployed on your CF?

**Comment by Callisto13** [comment]
So TLDR: when [Grootfs](https://github.com/cloudfoundry/grootfs) is called to report on a container's disk stats, it will return 2 values, the `total_bytes_used` and the `exclusive_bytes_used`. `exclusive` is the disk usage **not** including the rootfs image layers, and `total` is inclusive of those layers. Groot returns these to Garden, which returns them to the Diego Executor. Before Diego release 2.4.0, the executor parsed that json and returned the `exclusive_bytes_used` to the Cloud-Controller. Since 2.4.0 they return `total_bytes_used` which correctly accounts for the space taken up by the image and reports correctly to the CLI.


Pre Diego 2.4.0:
```sh
$ cf push matcha -o cloudfoundry/test-app # this app is ~23MB uncompressed
....

$ cf app matcha
Showing health and status for app match in org system / space system as admin...

name:              matcha
requested state:   started
instances:         1/1
usage:             256M x 1 instances
routes:            matcha.alienlifeform.garden-dev.cf-app.com
last uploaded:     Mon 09 Jul 15:02:55 BST 2018
stack:             cflinuxfs2
docker image:      cloudfoundry/test-app

     state     since                  cpu    memory         disk       details
#0   running   2018-07-09T14:03:12Z   0.2%   2.8M of 256M   8K of 1G
```
And on the diego-cell we can see where that 8K comes from...
```sh
$ ls /var/vcap/data/garden/depot/ # find the container handle
f526f6ce-4cb8-4966-4638-71b4
$ /var/vcap/packages/grootfs/bin/grootfs --config /var/vcap/jobs/garden/config/grootfs_config.yml stats f526f6ce-4cb8-4966-4638-71b4
{"disk_usage":{"total_bytes_used":23448093,"exclusive_bytes_used":8192}}
```

Diego 2.4.0 onwards:
```sh
$ cf push matcha -o cloudfoundry/test-app
....

$ cf app matcha
Showing health and status for app match in org system / space system as admin...

name:              matcha
requested state:   started
instances:         1/1
usage:             256M x 1 instances
routes:            matcha.ultrasonic-screwdriver.garden-dev.cf-app.com
last uploaded:     Tue 10 Jul 09:06:31 BST 2018
stack:             cflinuxfs2
docker image:      cloudfoundry/test-app

     state     since                  cpu    memory         disk          details
#0   running   2018-07-10T08:06:49Z   0.0%   2.7M of 256M   22.4M of 1G
```
And on the diego-cell...
```sh
$ ls /var/vcap/data/garden/depot/
55afbf65-5cbf-49c6-4461-f803
$ /var/vcap/packages/grootfs/bin/grootfs --config /var/vcap/jobs/garden/config/grootfs_config.yml stats 55afbf65-5cbf-49c6-4461-f803
{"disk_usage":{"total_bytes_used":23448093,"exclusive_bytes_used":8192}}
```
The correct number is reported.


I am closing this off now, please re-open if you see something different from above.

**Comment by scottgai** [comment]
@Callisto13

**Comment by teddyking** [comment]
For completeness - we followed up on this via slack and for OSS users the issue is resolved by upgrading to diego 2.4.0+.
